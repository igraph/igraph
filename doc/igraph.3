.\" Hey, Emacs!  This is an -*- nroff -*- source file.
.\"
.\" Copyright (C) 2006-2021  The igraph development team
.\"
.\" This is free software; you can redistribute it and/or modify it under
.\" the terms of the GNU General Public License as published by the Free
.\" Software Foundation; either version 2, or (at your option) any later
.\" version.
.\"
.\" This is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
.\" FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\" You should have received a copy of the GNU General Public License with
.\" your Debian GNU/Linux system, in /usr/share/common-licenses/GPL, or with
.\" the dpkg source package as the file COPYING.  If not, write to the Free
.\" Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH IGRAPH 3 "May 2021" "igraph library"
.SH NAME
igraph \- a library for creating and manipulating graphs
.SH DESCRIPTION
.B igraph
is a C library for complex network analysis and graph theory, with emphasis on
efficiency, portability and ease of use.
.SH DOCUMENTATION
The full documentation can be downloaded from the homepage of the
library:
.RI < https://igraph.org/c/doc >
.SH BUGS
If you think you have found a bug in igraph, feel free to file a bug report
in the issue tracker at:
.RI < https://github.com/igraph/igraph/issues >

.SH AUTHORS
Gabor Csardi <csardi.gabor@gmail.com>,
.br
Tamas Nepusz <ntamas@gmail.com>,
.br
Vincent Traag <vtraag@gmail.com>,
.br
Szabolcs Horvat <szhorvat@gmail.com>,
.br
Fabio Zanini <fabio.zanini@fastmail.fm>,
.br
Daniel Noom <ggatw@outlook.com>
