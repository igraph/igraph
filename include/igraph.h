/* -*- mode: C -*-  */
/*
   IGraph library.
   Copyright (C) 2003-2012  Gabor Csardi <csardi.gabor@gmail.com>
   334 Harvard street, Cambridge, MA 02139 USA

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc.,  51 Franklin Street, Fifth Floor, Boston, MA
   02110-1301 USA

*/

#ifndef IGRAPH_H
#define IGRAPH_H

#ifndef _GNU_SOURCE
    #define _GNU_SOURCE 1
#endif

#include "igraph_version.h"
#include "igraph_memory.h"
#include "igraph_error.h"
#include "igraph_random.h"
#include "igraph_progress.h"
#include "igraph_statusbar.h"

#include "igraph_types.h"
#include "igraph_complex.h"
#include "igraph_vector.h"
#include "igraph_matrix.h"
#include "igraph_array.h"
#include "igraph_dqueue.h"
#include "igraph_stack.h"
#include "igraph_heap.h"
#include "igraph_psumtree.h"
#include "igraph_strvector.h"
#include "igraph_vector_ptr.h"
#include "igraph_spmatrix.h"
#include "igraph_sparsemat.h"
#include "igraph_qsort.h"

#include "igraph_constants.h"
#include "igraph_datatype.h"
#include "igraph_iterators.h"
#include "igraph_interface.h"
#include "igraph_constructors.h"
#include "igraph_games.h"
#include "igraph_microscopic_update.h"
#include "igraph_centrality.h"
#include "igraph_paths.h"
#include "igraph_components.h"
#include "igraph_structural.h"
#include "igraph_transitivity.h"
#include "igraph_neighborhood.h"
#include "igraph_topology.h"
#include "igraph_bipartite.h"
#include "igraph_cliques.h"
#include "igraph_layout.h"
#include "igraph_visitor.h"
#include "igraph_community.h"
#include "igraph_conversion.h"
#include "igraph_foreign.h"
#include "igraph_motifs.h"
#include "igraph_operators.h"
#include "igraph_flow.h"
#include "igraph_nongraph.h"
#include "igraph_cocitation.h"
#include "igraph_adjlist.h"
#include "igraph_attributes.h"
#include "igraph_blas.h"
#include "igraph_lapack.h"
#include "igraph_arpack.h"
#include "igraph_mixing.h"
#include "igraph_separators.h"
#include "igraph_cohesive_blocks.h"
#include "igraph_eigen.h"
#include "igraph_hrg.h"
#include "igraph_threading.h"
#include "igraph_interrupt.h"
#include "igraph_scg.h"
#include "igraph_matching.h"
#include "igraph_embedding.h"
#include "igraph_scan.h"
#include "igraph_graphlets.h"
#include "igraph_epidemics.h"
#include "igraph_lsap.h"
#include "igraph_coloring.h"

#endif
