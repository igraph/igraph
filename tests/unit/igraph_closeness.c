#include <igraph.h>
#include <stdio.h>
#include "test_utilities.h"


void simple_test_case_no_weights_undirected(void) {

    igraph_t g;
    igraph_vector_t vector_actual_results;

    printf("Simple test case, no weights, undirected\n");

    igraph_vector_init(&vector_actual_results, 0);

    igraph_small(&g, 0, IGRAPH_DIRECTED, 0,1 , 1,2, -1);

    /* NOT NORMALISED TEST BELOW */

    igraph_closeness(&g, &vector_actual_results /*store results here*/,
                      NULL, NULL,
                      /*calculating for all vectors in the graph*/ igraph_vss_all(),
                      IGRAPH_ALL  /*graph is "undirected"*/,
                      NULL /*unweighted*/, /*not normalised*/ 0);

    printf("Non normalised results below\n");
    print_vector(&vector_actual_results);

    /* NORMALISED TEST BELOW */

    igraph_closeness(&g, &vector_actual_results /*store results here*/,
                      NULL, NULL,
                      /*calculating for all vectors in the graph*/ igraph_vss_all(),
                      IGRAPH_ALL  /*graph is "undirected"*/,
                      NULL, /*normalised*/ 1);

    printf("\nNormalised results below\n");
    print_vector(&vector_actual_results);

    igraph_vector_destroy(&vector_actual_results);
    igraph_destroy(&g);
}

void simple_test_case_with_weights_undirected(void) {

    igraph_t g;
    igraph_vector_int_t vector_edges;
    igraph_vector_t vector_weights, vector_actual_results;

    igraph_integer_t real_edges[] = {0,1 , 1,2};
    igraph_real_t real_weights[] = {3, 5};

    printf("\nSimple test case, with weights, undirected\n");

    igraph_vector_init(&vector_actual_results, 0);
    igraph_vector_int_view(&vector_edges, real_edges, sizeof(real_edges)/sizeof(real_edges[0]));
    igraph_create(&g, &vector_edges, /*number of vertices*/ 2, IGRAPH_DIRECTED);

    igraph_vector_view(&vector_weights, real_weights, sizeof(real_weights)/sizeof(real_weights[0]));

    /* NOT NORMALISED TEST BELOW */

    igraph_closeness(&g, &vector_actual_results /*store results here*/,
                      NULL, NULL,
                      /*calculating for all vectors in the graph*/ igraph_vss_all(),
                      IGRAPH_ALL  /*graph is "undirected"*/,
                      &vector_weights, /*not normalised*/ 0);

    printf("Non normalised test below\n");

    print_vector(&vector_actual_results);

    /* NORMALISED TEST BELOW */

    printf("\nNormalised test below\n");

    igraph_closeness(&g, &vector_actual_results /*store results here*/,
                      NULL, NULL,
                      /*calculating for all vectors in the graph*/ igraph_vss_all(),
                      IGRAPH_ALL  /*graph is "undirected"*/,
                      &vector_weights, /*normalised*/ 1);

    print_vector(&vector_actual_results);

    igraph_vector_destroy(&vector_actual_results);
    igraph_destroy(&g);
}

void advanced_test_case_no_weights_undirected(void) {

    igraph_t g;
    igraph_vector_t vector_actual_results;

    /* note, denominatory calculated as (shortest dist)*(n-1) for no normalisation
    normalisation excludes n-1 as part of the denominator */

    printf("\nAdvanced test case, no weights, undirected\n");

    igraph_vector_init(&vector_actual_results, 0);

    igraph_small(&g, 0, IGRAPH_DIRECTED, 1,0 , 0,5 , 5,6 , 5,4, 4,1 , 1,2 , 2,4 , 4,6 ,
                                 2,3 , 3,7 , 7,6 , 2,6 , -1);

    /* NOT NORMALISED TEST BELOW*/

    printf("Non normalised test below\n");

    igraph_closeness(&g, &vector_actual_results /*store results here*/,
                      NULL, NULL,
                      /*calculating for all vectors in the graph*/ igraph_vss_all(),
                      IGRAPH_ALL  /*graph is "undirected"*/,
                      NULL, /*not normalised*/ 0);

    print_vector(&vector_actual_results);

    /* NORMALISED TEST BELOW*/

    printf("\nNormalised test below\n");

    igraph_closeness(&g, &vector_actual_results /*store results here*/,
                      NULL, NULL,
                      /*calculating for all vectors in the graph*/ igraph_vss_all(),
                      IGRAPH_ALL  /*graph is "undirected"*/,
                      NULL, /*normalised*/ 1);

    print_vector(&vector_actual_results);

    igraph_vector_destroy(&vector_actual_results);
    igraph_destroy(&g);
}

void advanced_test_case_with_weights(void) {

    igraph_t g;
    igraph_vector_int_t vector_edges;
    igraph_vector_t vector_weights, vector_actual_results;

    igraph_integer_t real_edges[] = {1,0 , 0,5 , 5,6 , 5,4, 4,1 , 1,2 , 2,4 , 4,6 ,
                                 2,3 , 3,7 , 7,6 , 6,2};

    igraph_real_t real_weights[] = {4, 9, 2, 2, 2, 3, 1, 1, 8, 7, 5, 5};

    printf("\nAdvanced test case, with weights\n");


    igraph_vector_init(&vector_actual_results, 0);
    igraph_vector_int_view(&vector_edges, real_edges, sizeof(real_edges) / sizeof(real_edges[0]));
    igraph_create(&g, &vector_edges, /*number of vertices*/ 2, IGRAPH_DIRECTED);

    igraph_vector_view(&vector_weights, real_weights, sizeof(real_weights) / sizeof(real_weights[0]));

    /* TEST FOR UNDIRECTED GRAPH */

    printf("Undirected graph test below\n");

    igraph_closeness(&g, &vector_actual_results /*store results here*/,
                      NULL, NULL,
                      /*calculating for all vectors in the graph*/ igraph_vss_all(),
                      IGRAPH_ALL  /*graph is "undirected"*/,
                      &vector_weights, /*not normalised*/ 0);

    print_vector(&vector_actual_results);

    /* TEST FOR DIRECTED GRAPH
    OUT means the min distance from the curr node to the other node */

    printf("\nDirected graph test below for OUT\n");

    igraph_closeness(&g, &vector_actual_results /*store results here*/,
                      NULL, NULL,
                      /*calculating for all vectors in the graph*/ igraph_vss_all(),
                      IGRAPH_OUT  /*graph is "out directed"*/,
                      &vector_weights, /*not normalised*/ 0);

    print_vector(&vector_actual_results);

    /* IN means the min distance from a node to the curr node */

    printf("\nDirected graph test below for IN\n");

    igraph_closeness(&g, &vector_actual_results /*store results here*/,
                      NULL, NULL,
                      /*calculating for all vectors in the graph*/ igraph_vss_all(),
                      IGRAPH_IN  /*graph is "in directed"*/,
                      &vector_weights, /*not normalised*/ 0);

    print_vector(&vector_actual_results);

    igraph_vector_destroy(&vector_actual_results);
    igraph_destroy(&g);
}

void test_cutoff(void) {

    igraph_t g;
    igraph_vector_t closeness;
    igraph_vector_int_t reachable;
    igraph_bool_t all_reachable;
    size_t i;
    igraph_real_t cutoff_vec[] = { -1.0, 0.0, 1.0, 2.9, 3.0, 3.1 };

    printf("\n\nUnweighted undirected with cutoff\n");

    igraph_ring(&g, 4, IGRAPH_UNDIRECTED, 0, 0);

    igraph_vector_init(&closeness, 0);
    igraph_vector_int_init(&reachable, 0);

    for (i=0; i < sizeof(cutoff_vec) / sizeof(cutoff_vec[0]); ++i) {
        printf("\nRange-limited closeness with cutoff %g\n", cutoff_vec[i]);
        igraph_closeness_cutoff(&g, &closeness, &reachable, &all_reachable,
                                igraph_vss_all(), IGRAPH_ALL, NULL, /* normalized */ 1,
                                cutoff_vec[i]);
        printf("Closeness: ");
        print_vector(&closeness);
        printf("Reachable: ");
        print_vector_int(&reachable);
        printf("All reachable: %s\n", all_reachable ? "true" : "false");
    }

    igraph_vector_int_destroy(&reachable);
    igraph_vector_destroy(&closeness);

    igraph_destroy(&g);
}

void test_cutoff_directed(void) {

    igraph_t g;
    igraph_vector_t closeness;
    igraph_vector_int_t reachable;
    igraph_bool_t all_reachable;
    size_t i;
    igraph_real_t cutoff_vec[] = { -1.0, 0.0, 1.0, 2.9, 3.0, 3.1 };

    printf("\n\nUnweighted directed with cutoff\n");

    igraph_ring(&g, 4, IGRAPH_DIRECTED, 0, 0);

    igraph_vector_init(&closeness, 0);
    igraph_vector_int_init(&reachable, 0);

    for (i=0; i < sizeof(cutoff_vec) / sizeof(cutoff_vec[0]); ++i) {
        printf("\nRange-limited directed closeness with cutoff %g\n", cutoff_vec[i]);
        igraph_closeness_cutoff(&g, &closeness, &reachable, &all_reachable,
                                igraph_vss_all(), IGRAPH_OUT, NULL, /* normalized */ 1,
                                cutoff_vec[i]);
        printf("Closeness: ");
        print_vector(&closeness);
        printf("Reachable: ");
        print_vector_int(&reachable);
        printf("All reachable: %s\n", all_reachable ? "true" : "false");
    }

    igraph_vector_int_destroy(&reachable);
    igraph_vector_destroy(&closeness);

    igraph_destroy(&g);
}

void test_cutoff_weighted(void) {

    igraph_t g;
    igraph_vector_t closeness;
    igraph_vector_int_t reachable;
    igraph_bool_t all_reachable;
    size_t i;
    igraph_real_t cutoff_vec[] = { -1.0, 0.0, 1.0, 2.9, 3.0, 5.0, 6.0 };
    igraph_vector_t weights;

    printf("\n\nWeighted undirected with cutoff\n");

    igraph_ring(&g, 4, IGRAPH_UNDIRECTED, 0, 0);

    igraph_vector_init(&closeness, 0);
    igraph_vector_int_init(&reachable, 0);
    igraph_vector_init_range(&weights, 1, 4);

    for (i=0; i < sizeof(cutoff_vec) / sizeof(cutoff_vec[0]); ++i) {
        printf("\nRange-limited weighted closeness with cutoff %g\n", cutoff_vec[i]);
        igraph_closeness_cutoff(&g, &closeness, &reachable, &all_reachable,
                                igraph_vss_all(), IGRAPH_ALL, &weights, /* normalized */ 1,
                                cutoff_vec[i]);
        printf("Closeness: ");
        print_vector(&closeness);
        printf("Reachable: ");
        print_vector_int(&reachable);
        printf("All reachable: %s\n", all_reachable ? "true" : "false");
    }

    igraph_vector_destroy(&weights);
    igraph_vector_int_destroy(&reachable);
    igraph_vector_destroy(&closeness);

    igraph_destroy(&g);
}

void test_edge_cases(void) {

    igraph_t g;
    igraph_vector_t closeness;
    igraph_vector_int_t reachable;
    igraph_bool_t all_reachable;
    int n;

    printf("\n\nEdgeless graphs\n");

    igraph_vector_init(&closeness, 0);
    igraph_vector_int_init(&reachable, 0);

    for (n=0; n <= 2; ++n) {
        printf("\nEdgeless graph with %d vertices\n", n);
        igraph_empty(&g, n, IGRAPH_UNDIRECTED);

        igraph_closeness(&g, &closeness, &reachable, &all_reachable, igraph_vss_all(), IGRAPH_ALL, NULL, 1);
        printf("Closeness: ");
        print_vector(&closeness);
        printf("Reachable: ");
        print_vector_int(&reachable);
        printf("All reachable: %s\n", all_reachable ? "true" : "false");

        igraph_destroy(&g);
    }

    igraph_vector_int_destroy(&reachable);
    igraph_vector_destroy(&closeness);

}

int main(void) {

    simple_test_case_no_weights_undirected();
    simple_test_case_with_weights_undirected();
    advanced_test_case_no_weights_undirected();
    advanced_test_case_with_weights();

    test_cutoff();
    test_cutoff_directed();
    test_cutoff_weighted();

    test_edge_cases();

    VERIFY_FINALLY_STACK();

    return 0;
}
