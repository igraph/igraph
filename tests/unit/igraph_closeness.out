Simple test case, no weights, undirected
Non normalised results below
( 0.333333 0.5 0.333333 )

Normalised results below
( 0.666667 1 0.666667 )

Simple test case, with weights, undirected
Non normalised test below
( 0.0909091 0.125 0.0769231 )

Normalised test below
( 0.181818 0.25 0.153846 )

Advanced test case, no weights, undirected
Non normalised test below
( 0.0714286 0.0833333 0.1 0.0714286 0.1 0.0833333 0.1 0.0714286 )

Normalised test below
( 0.5 0.583333 0.7 0.5 0.7 0.583333 0.7 0.5 )

Advanced test case, with weights
Undirected graph test below
( 0.0169492 0.0285714 0.0322581 0.0140845 0.037037 0.027027 0.0333333 0.0192308 )

Directed graph test below for OUT
( 0.00869565 0.0172414 0.0192308 0.00763359 0.016129 0.0166667 0.0117647 0.01 )

Directed graph test below for IN
( 0.0128205 0.015873 0.015873 0.00980392 0.0188679 0.0075188 0.0263158 0.0075188 )


Unweighted undirected with cutoff

Range-limited closeness with cutoff -1
Closeness: ( 0.5 0.75 0.75 0.5 )
Reachable: ( 3 3 3 3 )
All reachable: true

Range-limited closeness with cutoff 0
Closeness: ( NaN NaN NaN NaN )
Reachable: ( 0 0 0 0 )
All reachable: false

Range-limited closeness with cutoff 1
Closeness: ( 1 1 1 1 )
Reachable: ( 1 2 2 1 )
All reachable: false

Range-limited closeness with cutoff 2.9
Closeness: ( 0.666667 0.75 0.75 0.666667 )
Reachable: ( 2 3 3 2 )
All reachable: false

Range-limited closeness with cutoff 3
Closeness: ( 0.5 0.75 0.75 0.5 )
Reachable: ( 3 3 3 3 )
All reachable: true

Range-limited closeness with cutoff 3.1
Closeness: ( 0.5 0.75 0.75 0.5 )
Reachable: ( 3 3 3 3 )
All reachable: true


Unweighted directed with cutoff

Range-limited directed closeness with cutoff -1
Closeness: ( 0.5 0.666667 1 NaN )
Reachable: ( 3 2 1 0 )
All reachable: false

Range-limited directed closeness with cutoff 0
Closeness: ( NaN NaN NaN NaN )
Reachable: ( 0 0 0 0 )
All reachable: false

Range-limited directed closeness with cutoff 1
Closeness: ( 1 1 1 NaN )
Reachable: ( 1 1 1 0 )
All reachable: false

Range-limited directed closeness with cutoff 2.9
Closeness: ( 0.666667 0.666667 1 NaN )
Reachable: ( 2 2 1 0 )
All reachable: false

Range-limited directed closeness with cutoff 3
Closeness: ( 0.5 0.666667 1 NaN )
Reachable: ( 3 2 1 0 )
All reachable: false

Range-limited directed closeness with cutoff 3.1
Closeness: ( 0.5 0.666667 1 NaN )
Reachable: ( 3 2 1 0 )
All reachable: false


Weighted undirected with cutoff

Range-limited weighted closeness with cutoff -1
Closeness: ( 0.3 0.375 0.375 0.214286 )
Reachable: ( 3 3 3 3 )
All reachable: true

Range-limited weighted closeness with cutoff 0
Closeness: ( NaN NaN NaN NaN )
Reachable: ( 0 0 0 0 )
All reachable: false

Range-limited weighted closeness with cutoff 1
Closeness: ( 1 1 NaN NaN )
Reachable: ( 1 1 0 0 )
All reachable: false

Range-limited weighted closeness with cutoff 2.9
Closeness: ( 1 0.666667 0.5 NaN )
Reachable: ( 1 2 1 0 )
All reachable: false

Range-limited weighted closeness with cutoff 3
Closeness: ( 0.5 0.666667 0.375 0.333333 )
Reachable: ( 2 2 3 1 )
All reachable: false

Range-limited weighted closeness with cutoff 5
Closeness: ( 0.5 0.375 0.375 0.25 )
Reachable: ( 2 3 3 2 )
All reachable: false

Range-limited weighted closeness with cutoff 6
Closeness: ( 0.3 0.375 0.375 0.214286 )
Reachable: ( 3 3 3 3 )
All reachable: true


Edgeless graphs

Edgeless graph with 0 vertices
Closeness: ( )
Reachable: ( )
All reachable: true

Edgeless graph with 1 vertices
Closeness: ( NaN )
Reachable: ( 0 )
All reachable: true

Edgeless graph with 2 vertices
Closeness: ( NaN NaN )
Reachable: ( 0 0 )
All reachable: false
